src/System.Private.CoreLib/src/System/MulticastDelegate.cs

```diff
@@ -135,32 +136,32 @@ public override sealed bool Equals(object obj)
         private bool InvocationListEquals(MulticastDelegate d)
         {
             Debug.Assert(d != null && (_invocationList as object[]) != null, "bogus delegate in multicast list comparison");
-            object[] invocationList = _invocationList as object[];
+            object[] invocationList = (_invocationList as object[])!; // Calling code ensures invocationList is object[]
             if (d._invocationCount != _invocationCount)
                 return false;
 
             int invocationCount = (int)_invocationCount;
             for (int i = 0; i < invocationCount; i++)
             {
-                Delegate dd = (Delegate)invocationList[i];
-                object[] dInvocationList = d._invocationList as object[];
+                Delegate dd = (Delegate)invocationList[i]!; // If invocationList is an object[], it always contains Delegate (or MulticastDelegate) objects
+                object[] dInvocationList = (d._invocationList as object[])!;
                 if (!dd.Equals(dInvocationList[i]))
                     return false;
             }
             return true;
         }
 
-        private bool TrySetSlot(object[] a, int index, object o)
+        private bool TrySetSlot(object?[] a, int index, object o)
         {
-            if (a[index] == null && System.Threading.Interlocked.CompareExchange<object>(ref a[index], o, null) == null)
+            if (a[index] == null && System.Threading.Interlocked.CompareExchange<object?>(ref a[index], o, null) == null)
                 return true;
 
             // The slot may be already set because we have added and removed the same method before.
             // Optimize this case, because it's cheaper than copying the array.
             if (a[index] != null)
             {
                 MulticastDelegate d = (MulticastDelegate)o;
-                MulticastDelegate dd = (MulticastDelegate)a[index];
+                MulticastDelegate dd = (MulticastDelegate)a[index]!;
```

Comment:
This looks like another case of the compiler bug that it doesn't see that we just validated `a[index] != null`.  Let's add the // TODO-NULLABLE comment here so that we can come back and clean it up when the compiler is fixed.
cc: @cston

Reply:
The compiler tracks the nullability of fields and properties of locals and parameters but does not track the nullability through indexers.

---
In reply to: [270860883](https://github.com/dotnet/coreclr/pull/23615#discussion_r270860883) [](ancestors = 270860883)

Reply:
Added comment.